testsuite/tests/typing-gadts/name_existentials.ml

```
@@ -84,7 +83,8 @@ let rec test : type a. a expr -> a = function
 Line 2, characters 22-23:
 2 |   | Int (type b) (n : a) -> n
                           ^
-Error: This type does not bind all existentials in the constructor: "type b. a"
+Error: This type annotation attempts to bind "b"
+       to the non-locally-abstract type "'a".
```

This error is confusing -- `'a` does not occur in the user code. Could it be improved? Does the `'a` come from the fact that `b` is not actually used in the pattern argument annotation? Maybe we can get a better message in this case.

---

I'm not sure how to track that `b` is not used other than by @alainfrisch 's logic for warnings, which I don't understand fully. Adding an extra level could work too, but this introduces new logic. Let me think about it.

---

I think you could actually check whether `b` is used in the pattern argument annotations without too much difficulty. For example, when you populate the initial environment with the `name_list` to get the `ids`, you could remember the identity of the flexible dummy variables for each `id`. Then on line 821 in the current PR, at the place where you raise the `Bind_non_locally_abstract` error, you can check whether the current identity of the type is still that flexible dummy variable, or if it was unified to something else. If it was not unified (it is still the dummy), you could raise a different error: the problem is not that the name is not being bound to a non-locally-abstract type, it is not being bound at all.

(I guess that this situation could occur in cases where the name does occur syntactically in the annotation, but those positions vanish during expansion. I don't think that this is a problem.)